File: blackout-kit-mobile/ios/Runner/VpnService.swift
import Flutter
import NetworkExtension

public class VpnServicePlugin: NSObject, FlutterPlugin {
  private var vpnConnection: VpnConnection?

  public static func dummy(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
    result(nil)
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "com.blackoutkit.vpn/service", binaryMessenger: registrar.messenger())
    let instance = VpnServicePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func dummyMethodToEnforceBundling() {
    // This method is intentionally empty
  }
}

// MARK: - FlutterPlugin Method Handler
extension VpnServicePlugin {
  public func methodCallHandler(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "connect":
      handleConnect(call, result: result)
    case "disconnect":
      handleDisconnect(result: result)
    case "isRunning":
      result(vpnConnection?.isRunning ?? false)
    case "getStatus":
      handleGetStatus(result: result)
    case "getConnectedIP":
      handleGetConnectedIP(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleConnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "Arguments are null", details: nil))
      return
    }

    guard let protocol = args["protocol"] as? String,
          let address = args["address"] as? String,
          let port = args["port"] as? NSNumber else {
      result(FlutterError(code: "MISSING_ARGS", message: "Missing required arguments", details: nil))
      return
    }

    let displayName = (args["displayName"] as? String) ?? "VPN"
    let rawUri = (args["rawUri"] as? String) ?? ""

    vpnConnection = VpnConnection(
      protocol: `protocol`,
      address: address,
      port: port.intValue,
      displayName: displayName,
      rawUri: rawUri
    )

    // Protocol-specific configuration
    switch `protocol` {
    case "wireguard":
      if let privateKey = args["privateKey"] as? String {
        vpnConnection?.privateKey = privateKey
      }

    case "openvpn":
      if let configContent = args["configContent"] as? String {
        vpnConnection?.configContent = configContent
      }

    case "shadowsocks":
      if let method = args["method"] as? String {
        vpnConnection?.method = method
      }
      if let password = args["password"] as? String {
        vpnConnection?.password = password
      }
      if let plugin = args["plugin"] as? String {
        vpnConnection?.plugin = plugin
      }

    default:
      result(FlutterError(code: "UNSUPPORTED_PROTOCOL", message: "Unsupported protocol: \(`protocol`)", details: nil))
      return
    }

    vpnConnection?.connect()
    result(["success": true, "message": "Connected"])
  }

  private func handleDisconnect(result: @escaping FlutterResult) {
    vpnConnection?.disconnect()
    vpnConnection = nil
    result(["success": true])
  }

  private func handleGetStatus(result: @escaping FlutterResult) {
    let status: [String: Any] = [
      "isConnected": vpnConnection?.isConnected ?? false,
      "isRunning": vpnConnection?.isRunning ?? false,
      "protocol": vpnConnection?.protocol ?? "unknown",
      "displayName": vpnConnection?.displayName ?? "",
      "address": vpnConnection?.address ?? ""
    ]
    result(status)
  }

  private func handleGetConnectedIP(result: @escaping FlutterResult) {
    result(vpnConnection?.getConnectedIP() ?? "Unknown")
  }
}

// MARK: - VpnConnection
class VpnConnection {
  let `protocol`: String
  let address: String
  let port: Int
  let displayName: String
  let rawUri: String

  var privateKey: String?
  var configContent: String?
  var method: String?
  var password: String?
  var plugin: String?

  var isConnected: Bool = false
  var isRunning: Bool = false
  private var connectedIP: String = "0.0.0.0"

  init(protocol: String, address: String, port: Int, displayName: String, rawUri: String) {
    self.protocol = `protocol`
    self.address = address
    self.port = port
    self.displayName = displayName
    self.rawUri = rawUri
  }

  func connect() {
    isConnected = true
    isRunning = true
    connectedIP = generateMockIP()
    print("VpnConnection: Connected to \(displayName) via \(`protocol`)")
  }

  func disconnect() {
    isConnected = false
    isRunning = false
    connectedIP = "0.0.0.0"
    print("VpnConnection: Disconnected from \(displayName)")
  }

  func getConnectedIP() -> String {
    return isConnected ? connectedIP : "0.0.0.0"
  }

  private func generateMockIP() -> String {
    let random = Int.random(in: 1...254)
    return "192.168.1.\(random)"
  }
}

File: android/app/src/main/kotlin/com/blackoutkit/vpn/VpnService.kt
package com.blackoutkit.vpn

import android.app.Service
import android.content.Intent
import android.net.VpnService
import android.os.Binder
import android.os.IBinder
import android.os.ParcelFileDescriptor
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodCall
import android.util.Log

/// VPN Service for Blackout Kit - manages VPN connection on Android
class VpnServicePlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var vpnConnection: VpnConnection? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "connect" -> handleConnect(call, result)
            "disconnect" -> handleDisconnect(result)
            "isRunning" -> result.success(vpnConnection?.isRunning ?: false)
            "getStatus" -> handleGetStatus(result)
            "getConnectedIP" -> handleGetConnectedIP(result)
            else -> result.notImplemented()
        }
    }

    private fun handleConnect(call: MethodCall, result: Result) {
        try {
            val protocol = call.argument<String>("protocol") ?: return
            val address = call.argument<String>("address") ?: return
            val port = call.argument<Int>("port") ?: return
            val displayName = call.argument<String>("displayName") ?: "VPN"
            val rawUri = call.argument<String>("rawUri") ?: ""

            vpnConnection = VpnConnection(
                protocol = protocol,
                address = address,
                port = port,
                displayName = displayName,
                rawUri = rawUri
            )

            // For WireGuard
            if (protocol == "wireguard") {
                val privateKey = call.argument<String>("privateKey") ?: ""
                vpnConnection?.privateKey = privateKey
            }

            // For OpenVPN
            if (protocol == "openvpn") {
                val configContent = call.argument<String>("configContent") ?: ""
                vpnConnection?.configContent = configContent
            }

            // For Shadowsocks
            if (protocol == "shadowsocks") {
                val method = call.argument<String>("method") ?: "aes-256-gcm"
                val password = call.argument<String>("password") ?: ""
                val plugin = call.argument<String>("plugin")
                vpnConnection?.method = method
                vpnConnection?.password = password
                vpnConnection?.plugin = plugin
            }

            vpnConnection?.connect()
            result.success(mapOf("success" to true, "message" to "Connected"))
        } catch (e: Exception) {
            Log.e(TAG, "Connection error", e)
            result.error("CONNECT_ERROR", e.message, null)
        }
    }

    private fun handleDisconnect(result: Result) {
        try {
            vpnConnection?.disconnect()
            vpnConnection = null
            result.success(mapOf("success" to true))
        } catch (e: Exception) {
            Log.e(TAG, "Disconnect error", e)
            result.error("DISCONNECT_ERROR", e.message, null)
        }
    }

    private fun handleGetStatus(result: Result) {
        val status = mapOf(
            "isConnected" to (vpnConnection?.isConnected ?: false),
            "isRunning" to (vpnConnection?.isRunning ?: false),
            "protocol" to (vpnConnection?.protocol ?: "unknown"),
            "displayName" to (vpnConnection?.displayName ?: ""),
            "address" to (vpnConnection?.address ?: "")
        )
        result.success(status)
    }

    private fun handleGetConnectedIP(result: Result) {
        try {
            val ip = vpnConnection?.getConnectedIP() ?: "Unknown"
            result.success(ip)
        } catch (e: Exception) {
            result.error("IP_ERROR", e.message, null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivity() {}
    override fun onReattachedToActivity(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivityForConfigChanges() {}
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}

    companion object {
        private const val CHANNEL = "com.blackoutkit.vpn/service"
        private const val TAG = "VpnServicePlugin"
    }
}

/// VPN Connection - represents active VPN connection
class VpnConnection(
    val protocol: String,
    val address: String,
    val port: Int,
    val displayName: String,
    val rawUri: String
) {
    var privateKey: String? = null
    var configContent: String? = null
    var method: String? = null
    var password: String? = null
    var plugin: String? = null

    var isConnected = false
    var isRunning = false
    private var connectedIP: String = "0.0.0.0"

    fun connect() {
        // Simulate connection
        isConnected = true
        isRunning = true
        connectedIP = generateMockIP()
        Log.i("VpnConnection", "Connected to $displayName via $protocol")
    }

    fun disconnect() {
        isConnected = false
        isRunning = false
        connectedIP = "0.0.0.0"
        Log.i("VpnConnection", "Disconnected from $displayName")
    }

    fun getConnectedIP(): String {
        return if (isConnected) connectedIP else "0.0.0.0"
    }

    private fun generateMockIP(): String {
        // Mock: Return a different IP each time (would be actual IP in real implementation)
        val random = (1..254).random()
        return "192.168.1.$random"
    }
}
